bpf:nat: restore a NAT entry if its REV NAT is not found

sugangli opened the pull request:
<!-- Description of change -->
NAT Map LRU behavior could evict any entry in the MAP. This change restore a reverse NAT entry when it is expected but not found.  Previously, we were only relying on existence of forwarding NAT to determine if the NAT pair exists or not.  

This fix is verified by using the same repro steps in  #34833.  After the fix, we saw much less timeout from the client, but the case like eviction happens between TCP SYN and TCP SYN ACK will still drop the returning packets.

Fixes: #34833

sugangli committed:
bpf:nat: restore a NAT entry if its REV NAT is not found

Previously, we were only relying on existence of forwarding NAT
to determine if the NAT pair exists or not. NAT MAP's LRU eviction
could evict the reverse NAT entry and cause return traffic gettting
dropped.

Signed-off-by: Sugang Li <sugangli@google.com>

sypakine commented on bpf/lib/nat.h:
```diff
@@ -312,8 +313,10 @@ snat_v4_nat_handle_mapping(struct __ctx_buff *ctx,
-		if (target->addr == (*state)->to_saddr &&
+		set_v4_rtuple(tuple, *state, &rtuple);
+		rstate = __snat_lookup(map, &rtuple);
+		if (rstate &&
```
Add a comment explaining that we want to ensure there is a reverse entry.

sypakine commented on bpf/lib/nat.h:
```diff
@@ -312,24 +313,25 @@ snat_v4_nat_handle_mapping(struct __ctx_buff *ctx,
+		    target->addr == (*state)->to_saddr &&
 		    needs_ct == (*state)->common.needs_ct) {
 			barrier_data(*state);
 			return 0;
```
Aside: should this be replaced with `CTX_ACT_OK`?

sugangli replied:
This is not the final stage of the packet. Returning CTX_ACT_OK may be too early?

sypakine replied:
`CTX_ACT_OK` is just an alias for `0`

Don't worry about this though, marked it as an aside (i.e. a follow-up for anyone).

sugangli replied:
Yup.  I was checking  the convention of  using `CTX_ACT_OK`,  other places seem to be the final stages.

sypakine commented on bpf/lib/nat.h:
```diff
@@ -312,24 +313,25 @@ snat_v4_nat_handle_mapping(struct __ctx_buff *ctx,
 		struct ipv4_ct_tuple rtuple = {};
-
-		if (target->addr == (*state)->to_saddr &&
+		/* reverse SNAT entries could be missing due to LRU eviction */
```
nit: this comment could be clearer:

```
Check for the reverse SNAT entry. If it is missing (e.g. due to LRU eviction), it must be restored before returning.
```

sugangli replied:
Done.

sypakine commented on bpf/lib/nat.h:
```diff
@@ -312,24 +313,25 @@ snat_v4_nat_handle_mapping(struct __ctx_buff *ctx,
-		/* Recreate the SNAT and RevSNAT entries if the source IP is stale.
-		 * Otherwise, the packet will be erroneously SNATed with the stale
-		 * source IP.
+		/* Recreate the SNAT and RevSNAT entries if the source IP is stale,
```
consider: when there are multiple cases, it reads better as a list and allows items to be added without touching other comment lines unnecessarily:

```
Recreate the SNAT and RevSNAT entries if any of:
- Source IP is stale
- RevSNAT entry is missing

Otherwise, the packet will be erroneously SNATed with the stale source IP or dropped.
```

sugangli replied:
Done.

tommyp1ckles commented on bpf/lib/nat.h:
```diff
@@ -312,24 +313,28 @@ snat_v4_nat_handle_mapping(struct __ctx_buff *ctx,
 		struct ipv4_ct_tuple rtuple = {};
-
-		if (target->addr == (*state)->to_saddr &&
+		/* Check for the reverse SNAT entry. If it is missing (e.g. due to LRU
```
Changes seem reasonable, I'm a little bit concerned with any possible regressions here resulting in unnecessary/unnoticed recreations of nat mappings.

Perhaps it would make sense to make sure there's some coverage of this particular scenario over in bpf/tests/bpf_nat_tests.c?

sugangli replied:
@tommyp1ckles Thanks for the comment.  I was thinking what could be an unnoticed recreation? I am assuming fwd tuple and rev tuple should always co-exist. The only case is when they become orphaned, we don't need to recreate them and it is OK to stay broken. But this path won't touch the orphaned entries.

sugangli replied:
Added the tests.

julianwiedmann reviewed (changes requested):
Could you please also add the IPv6 part of the fix?

Agree with Tom that this needs tests. I could see a `bpf_host` test under `bpf/tests/` that walks through the desired packet flow, and reproduces the whole scenario you're fixing:

1. transmit a packet, have it SNATed
2. receive a reply for that connection, check that it gets revSNATed
3. manually delete the revSNAT entry
4. receive a second reply, check that it no longer gets revSNATed 
5. transmit a second packet, validate that it re-creates the missing state
6. receive a third reply, check that it gets revSNATed again

sugangli committed:
bpf:tests: Add tests to cover restored Rev NAT case

The following three tests are checking the scenario where a Rev NAT entry gets
deleleted:

1. The first reply gets punted to the kernel as its NAT lookup fails.
2. The following request gets to the backend and restores the Rev NAT.
3. The second reply gets back to the client normally.

Signed-off-by: Sugang Li <sugangli@google.com>

tommyp1ckles commented on bpf/tests/tc_nodeport_lb4_nat_lb.c:
```diff
@@ -914,3 +914,212 @@ int nodeport_nat_fwd_reply_no_fib_check(__maybe_unused const struct __ctx_buff *
 }
+
+/* The following three tests are checking the scenario where a Rev NAT entry gets
+ * deleleted:
```
typo: deleleted

sugangli replied:
Done.

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1220,9 +1231,7 @@ snat_v6_nat_handle_mapping(struct __ctx_buff *ctx,
-		set_v6_rtuple(tuple, *state, &rtuple);
-		*state = snat_v6_lookup(&rtuple);
-		if (*state)
+		if (rstate)
```
(commenting here as the diff is cleaner)

to clarify - when the RevSNAT entry is missing, we
1. delete the SNAT entry
2. let `snat_v6_new_mapping()` create a new SNAT / RevSNAT pair of entries.

In a busy node, won't `snat_v6_new_mapping()` then end up selecting a different source port for this connection? Won't this result in disrupted connectivity?

I would have expected that we preserve the SNAT entry, and merely re-create the RevSNAT entry from the available information. Why is that not possible?

sugangli replied:
Good point. Updated!

julianwiedmann commented on bpf/tests/tc_nodeport_lb4_nat_lb.c:
```diff
@@ -914,3 +914,212 @@ int nodeport_nat_fwd_reply_no_fib_check(__maybe_unused const struct __ctx_buff *
+ * Test that the LB fails to RevDNAT and RevSNAT a reply from the
+ * NAT remote backend when its Rev NAT entry gets deleted
+ */
+PKTGEN("tc", "tc_nodeport_nat_fwd_reply_punt")
+int nodeport_nat_fwd_reply_punt_pktgen(struct __ctx_buff *ctx)
+{
+	return build_reply(ctx);
+}
```
Thank you, the test looks good!

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -313,8 +313,24 @@ snat_v4_nat_handle_mapping(struct __ctx_buff *ctx,
+				rstate.to_dport = tuple->sport;
+				rstate.common.needs_ct = needs_ct;
+				rstate.common.created = bpf_mono_now();
+				return __snat_create(map, &rtuple, &rstate);
```
Please align this with how we handle errors from `__snat_create()` in other places (setting `ext_err`, and returning a proper DROP reason)

sugangli replied:
Done.

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -327,8 +343,6 @@ snat_v4_nat_handle_mapping(struct __ctx_buff *ctx,
 			return ret;
 
-		set_v4_rtuple(tuple, *state, &rtuple);
-		*state = __snat_lookup(map, &rtuple);
```
Can you explain why this lookup is no longer needed :pray: ? It's not obvious to me.

sugangli replied:
I was just reading the comment below that it was OK to fail a deletion? If that's the case, maybe we can always try to delete RevSNAT if the SNAT exists.  But I am not sure about the penalty of deletion failure. Adding it back.

julianwiedmann reviewed (approved):
lgtm, thank you! I still feel this is a bit hack-ish, and hope we can find a solution to this issue that covers the whole problem space.
